cpp/src/arrow/array/array_primitive.cc

```diff
@@ -56,7 +56,7 @@ int64_t BooleanArray::false_count() const {
 }
 
 int64_t BooleanArray::true_count() const {
-  if (data_->null_count.load() != 0) {
+  if (data_->GetNullCount() != 0) {
```

Comment:
so the problem is that validity buffer might not exists, which might cause segment fault here?

Reply:
Yes (in general, I think we should never use ArrayData.null_count directly, unless in some low level code that deals directly with that, like the implementation of `GetNullCount`)

Reply:
Will approve after https://github.com/apache/arrow/pull/41070#discussion_r1556403029 fixed

By the way, should this be in 16.0?